### PR TITLE
Split out advanced Blazor hosting model content

### LINKED ISSUE #16764: Split advanced configuration content out of Blazor Hosting Models topic
For discussion/triage: The *Hosting Models* topic goes into advanced configuration details that might be better off in another topic.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 542f38c4-1502-f739-372a-80c86c377491
* Version Independent ID: 72c2a844-ffb2-4799-2c7a-a7a73d3b6833
* Content: [ASP.NET Core Blazor hosting models](https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-3.1)
* Content Source: [aspnetcore/blazor/hosting-models.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/hosting-models.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #16860: Clarification about OnInitializedAsync in a pre-rendered Blazor Server app
The comment [here](https://docs.microsoft.com/en-us/aspnet/core/blazor/lifecycle?view=aspnetcore-3.1#lifecycle-methods) states "These methods are only called one time when the component is first instantiated." This is not actually true in a Pre-rendered scenario, which is the default for Blazor Server. Therefore, people may inadvertently put initialisation logic such as data loading, or delegate hooks, into this method, and not realise that it gets called twice.

It might make sense to put an info box with the caveat explaining that it can get called twice, and link to [this page](https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-3.1#stateful-reconnection-after-prerendering) - and it may be useful to have a full example explaining an approach to loading initialisation data in this context.
See this issue for background: https://github.com/dotnet/aspnetcore/issues/18815

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 4bae853d-80e3-ef81-69eb-e8276eb853fe
* Version Independent ID: 01216f14-b45c-73b0-9f1f-300e7ddd788f
* Content: [ASP.NET Core Blazor lifecycle](https://docs.microsoft.com/en-us/aspnet/core/blazor/lifecycle?view=aspnetcore-3.1#lifecycle-methods)
* Content Source: [aspnetcore/blazor/lifecycle.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/lifecycle.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #16764
Fixes #16860

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/hosting-model-configuration?view=aspnetcore-3.1&branch=pr-en-us-16845)

* The current hosting models topic bogs down new-to-Blazor readers with advanced how-to-configure content. This PR takes the topic back to where we started ... primary concepts of hosting models. This frees the reader to get on to the build-your-first experience when consuming these topics in a linear fashion. The content is moved out to a new topic, *Blazor hosting model configuration*.
* Let's leave the unpaired 2nd heading level for *Blazor Server* and just comment out the *Blazor WebAssembly* heading for now. I think we'll have some hosting model config for BW later.
* In the Blazor TOC, let's consider moving the layouts topic up a bit and the component library topic down a bit.
* Move Blazor Server prerendering content into the *Lifecycle* topic. However, let's not shoehorn the content into the *Component initialization methods* (`OnInitializedAsync`) section at the top. I place the coverage in sections at the bottom of the topic and cross-link it from the *Component initialization methods* section. In this way, the reader isn't distracted from the focus on the lifecycle methods. Thanks @Webreaper, who reported the confusing layout of coverage.

:question: **_Naming_**: Do you prefer a different name/filename/URL than:

* *Blazor hosting model configuration*
* *blazor/hosting-model-configuration.md*

### CODE REVIEW

File: aspnetcore/blazor/hosting-model-configuration.md [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+     ```html
+     <base href="~/" />
+     ```
+
+     The `href` value (the *app base path*) in the preceding example assumes that the app resides at the root URL path (`/`). If the app is a sub-application, follow the guidance in the *App base path* section of the <xref:host-and-deploy/blazor/index#app-base-path> article.
+
+     The *_Layout.cshtml* file is located in the *Pages/Shared* folder in a Razor Pages app or *Views/Shared* folder in an MVC app.
+
+   * Add a `<script>` tag for the *blazor.server.js* script inside of the closing `</body>` tag:
```
Comment (mkArtak) [inline aspnetcore/blazor/hosting-model-configuration.md]:
```suggestion
   * Add a `<script>` tag for the *blazor.server.js* script right before of the closing `</body>` tag:
```

File: aspnetcore/blazor/hosting-model-configuration.md [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+   @using Microsoft.AspNetCore.Authorization
+   @using Microsoft.AspNetCore.Components.Authorization
+   @using Microsoft.AspNetCore.Components.Forms
+   @using Microsoft.AspNetCore.Components.Routing
+   @using Microsoft.AspNetCore.Components.Web
+   @using Microsoft.JSInterop
+   @using MyAppNamespace
+   ```
+
+1. In `Startup.ConfigureServices`, add the Blazor Server service:
```
Comment (mkArtak) [inline aspnetcore/blazor/hosting-model-configuration.md]:
```suggestion
1. In `Startup.ConfigureServices`, add registration for the Blazor Server service:
```

File: aspnetcore/blazor/hosting-model-configuration.md [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+      <component type="typeof(App)" render-mode="ServerPrerendered" />
+    </app>
+
+    <script src="_framework/blazor.server.js"></script>
+</body>
+```
+
+`RenderMode` configures whether the component:
+
+* Is prerendered into the page.
```
Comment (mkArtak) [inline aspnetcore/blazor/hosting-model-configuration.md]: Do we have an article which actually details out what `prerendering` is?
As a non-native English speaker, it was a bit unclear to me what `prerendering` is. Although I now know it, I think it's still valuable to have a writeup about this - "what is prerendering actually".
Looking at these two sentences

- Is prerendered into the page.
- Is rendered as static HTML on the page


it's not clear with understanding of what prerendering is, what's the actual difference.

Comment (guardrex) [reply aspnetcore/blazor/hosting-model-configuration.md]: This topic is the first place where it appears, so I'll define it here before merging.

**Pull request merged**